### PR TITLE
Add rtti package

### PR DESCRIPTION
**Pull request by qwqllh**

### CODE REVIEW

File: packages/r/rttr/xmake.lua [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        add_versions("0.9.5", "bab810d99d7c8191fde0032061d6120e4bc99eebb239036e2adaf34ebe92a51f")
+    else
+        add_urls("https://www.rttr.org/releases/rttr-$(version)-src.tar.gz",
+                 "https://github.com/rttrorg/rttr/releases/download/v$(version)/rttr-$(version)-src.tar.gz",
+                 "https://github.com/rttrorg/rttr.git")
+
+        -- unix version use \n
+        add_versions("0.9.6", "f62caee43016489320f8a69145c9208cddd72e451ea95618bc26a49a4cd6c990")
+        add_versions("0.9.5", "caa8d404840b0e156f869a947e475b09f7b602ab53c290271f40ce028c8d7d91")
+    end
```
Comment (waruqi) [inline packages/r/rttr/xmake.lua:26]: why? please use same url.

it should work for `\n` for win/linux

File: packages/r/rttr/xmake.lua [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+    on_install(function (package)
+        local configs = {}
+        table.insert(configs, "-DCMAKE_BUILD_TYPE=" .. (package:debug() and "Debug" or "Release"))
```
Comment (waruqi) [inline packages/r/rttr/xmake.lua:30]: please add static/shared kind
```
        table.insert(configs, "-DBUILD_SHARED_LIBS=" .. (package:config("shared") and "ON" or "OFF"))
```

File: packages/r/rttr/xmake.lua [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+        table.insert(configs, "-DBUILD_EXAMPLES=OFF")
+        table.insert(configs, "-DBUILD_DOCUMENTATION=OFF")
+        table.insert(configs, "-DBUILD_UNIT_TESTS=OFF") -- rttr has problem building unit tests on macosx.
+        import("package.tools.cmake").install(package, configs)
+    end)
```
Comment (waruqi) [inline packages/r/rttr/xmake.lua:35]: please add `on_test`

File: packages/r/rttr/xmake.lua (hunk 1 above)
Comment (qwqllh) [reply packages/r/rttr/xmake.lua:26]: The package maintainer offeres different versions for windows and unix. I'll modify this later.

File: packages/r/rttr/xmake.lua [hunk 4]
```diff
@@ -0,0 +1,76 @@
+package("rttr")
+    set_kind("shared")
```
Comment (waruqi) [inline packages/r/rttr/xmake.lua:2]: please remove it

File: packages/r/rttr/xmake.lua [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    set_homepage("https://www.rttr.org")
+    set_description("rttr: An open source library, which adds reflection to C++.")
```
Comment (waruqi) [inline packages/r/rttr/xmake.lua:4]: set_license

File: packages/r/rttr/xmake.lua [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        add_extsources("brew::rttr")
+    end
+    
+    add_urls("https://www.rttr.org/releases/rttr-$(version)-src.tar.gz",
+             "https://github.com/rttrorg/rttr/releases/download/v$(version)/rttr-$(version)-src.tar.gz",
+             "https://github.com/rttrorg/rttr.git")
+
+    add_versions("0.9.6", "f62caee43016489320f8a69145c9208cddd72e451ea95618bc26a49a4cd6c990")
+    add_versions("0.9.5", "caa8d404840b0e156f869a947e475b09f7b602ab53c290271f40ce028c8d7d91")
+
```
Comment (waruqi) [inline packages/r/rttr/xmake.lua:16]: add_deps("cmake")

**Comment by qwqllh** [comment]
Thanks for your advice. But I'm currently working on my mid-term examination nowadays and I do not have a PC to fix problems on Windows. Maybe I will work on this later.

File: packages/r/rttr/xmake.lua [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    on_install(function (package)
+        local configs = {}
+        table.insert(configs, "-DCMAKE_BUILD_TYPE=" .. (package:debug() and "Debug" or "Release"))
+        table.insert(configs, "-DBUILD_EXAMPLES=OFF")
+        table.insert(configs, "-DBUILD_DOCUMENTATION=OFF")
+        table.insert(configs, "-DBUILD_UNIT_TESTS=OFF") -- rttr has problem building unit tests on macosx.
+        -- rttr use BUILD_RTTR_DYNAMIC and BUILD_STATIC options to control whether to build dynamic or static libraries.
+        table.insert(configs, "-DBUILD_RTTR_DYNAMIC=" .. (package:config("shared") and "ON" or "OFF"))
+        table.insert(configs, "-DBUILD_STATIC=" .. (package:config("shared") and "OFF" or "ON"))
+        import("package.tools.cmake").install(package, configs, {cxxflags = "-Wno-error=implicit-const-int-float-conversion"})
```
Comment (waruqi) [inline packages/r/rttr/xmake.lua:28]: only for gcc/clang 

if package:has_tool("cxx", "gcc", "gxx", "clang", "clangxx") then

**Comment by qwqllh** [comment]
For shared test on windows, I find that the major reason for linking error may be the rttr shared library install path configuration since CMAKE_INSTALL_LIBDIR is not used on Windows. Is it possible to specify the library path in xmake.lua manually?

**Comment by waruqi** [comment]
> For shared test on windows, I find that the major reason for linking error may be the rttr shared library install path configuration since CMAKE_INSTALL_LIBDIR is not used on Windows. Is it possible to specify the library path in xmake.lua manually?

xmake has added it. https://github.com/xmake-io/xmake/blob/564876e408be47c73644fb9a060790c71f45e8f6/xmake/modules/package/tools/cmake.lua#L520

**Comment by waruqi** [comment]
maybe it does not export symbols, you can try use `-DCMAKE_WINDOWS_EXPORT_ALL_SYMBOLS=ON` or other macro switch for this project.

**Comment by qwqllh** [comment]
Its native unit test works well on windows with shared library when building with cmake manually. I don't think that's because symbols. I mean that the CMAKE_INSTALL_LIBDIR variable is unused when building on windows. The install path is hard coded like this:

```cmake
# set all install directories for the targets
if(UNIX)
  include(GNUInstallDirs)
  set(RTTR_RUNTIME_INSTALL_DIR "${CMAKE_INSTALL_BINDIR}") 
  set(RTTR_LIBRARY_INSTALL_DIR "${CMAKE_INSTALL_LIBDIR}")
  set(RTTR_ARCHIVE_INSTALL_DIR "${CMAKE_INSTALL_LIBDIR}")
  set(RTTR_FRAMEWORK_INSTALL_DIR "${CMAKE_INSTALL_LIBDIR}")

  set(RTTR_INSTALL_FULL_LIBDIR "${CMAKE_INSTALL_FULL_LIBDIR}")

  set(RTTR_CMAKE_CONFIG_INSTALL_DIR "${CMAKE_INSTALL_DATADIR}/rttr/cmake")
  set(RTTR_ADDITIONAL_FILES_INSTALL_DIR "${CMAKE_INSTALL_DATADIR}/rttr")

elseif(WIN32)
  set(RTTR_RUNTIME_INSTALL_DIR   "bin") 
  set(RTTR_LIBRARY_INSTALL_DIR   "bin")
  set(RTTR_ARCHIVE_INSTALL_DIR   "lib")
  set(RTTR_FRAMEWORK_INSTALL_DIR "bin")

  set(RTTR_CMAKE_CONFIG_INSTALL_DIR "${CMAKE_INSTALL_PREFIX}/cmake")
  set(RTTR_ADDITIONAL_FILES_INSTALL_DIR "${CMAKE_INSTALL_PREFIX}")
else()
  message(FATAL_ERROR "Could not set install folders for this platform!")
endif()
```

I tried to fix this with a patch. But I haven't figured out how to set the RTTR_LIBRARY_INSTALL_DIR currectly yet.

**Comment by waruqi** [comment]
you need define RTTR_DLL to define export symbols for on_test. I tested it, it works for me.

```lua
        if package:is_plat("windows") and package:config("shared") then
            package:add("defines", "RTTR_DLL")
        end
        import("package.tools.cmake").install(package, configs, {cxflags = extra_cxflags})
```

https://github.com/rttrorg/rttr/blob/7edbd580cfad509a3253c733e70144e36f02ecd4/src/rttr/detail/base/core_prerequisites.h#L137

https://github.com/rttrorg/rttr/blob/b16fccf0fbbbf94064bf2a6c7c47f2b910ab31f1/src/rttr/CMakeLists.txt#L47

**Pull request merged**